test/inductor/test_torchinductor_dynamic_shapes.py

```diff
@@ -609,8 +610,92 @@ def func(x, fn, a):
+                return x.cos()
+            if x.shape[0] < 5:
+                return x * 2
+            # x.shape[0] == 5 at this point
+            return x.sin()
+
+        def fn_3(x):
+            # equality constraint, which matches example shape
+            if x.size(0) == 5:
+                return x.sin()
+            else:
+                return x.cos()
+
+        call_count = 0
+
+        def _test_wrapper_codegen_statically_known_int_or_none_in_context():
+            nonlocal call_count
+            call_count += 1
+            graph = V.graph
+            input_layouts = [
+                inp.layout
+                for inp in graph.graph_inputs.values()
+                if hasattr(inp, "layout")
+            ]
+            batch_dim = input_layouts[0].size[0]
+            if call_count == 1:
+                # testing fn_1
+                assert (
+                    WrapperCodeGen.statically_known_int_or_none(batch_dim) is None
+                ), "Should not be statically known on first call"
+            elif call_count == 2:
+                # testing fn_2
+                assert (
+                    WrapperCodeGen.statically_known_int_or_none(batch_dim) == 5
+                ), "Should be limited to exactly 5 on third call due to multiple constraints"
+            elif call_count == 2:
+                # testing fn_3
+                assert (
+                    WrapperCodeGen.statically_known_int_or_none(batch_dim) == 5
+                ), "Should be exactly 5 on second call"
+
+        class TestWrapperCodegen(WrapperCodeGen):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, **kwargs)
+
+            def generate(self, is_inference, *args, **kwargs):
+                _test_wrapper_codegen_statically_known_int_or_none_in_context()
+                return super().generate(is_inference, *args, **kwargs)
+
+        if "cpu" not in device_codegens:
+            register_backend_for_device("cpu", CppScheduling, WrapperCodeGen)
+        orig_cpu_codegens = device_codegens["cpu"]
+        try:
+            register_backend_for_device(
+                "cpu", orig_cpu_codegens.scheduling, TestWrapperCodegen
+            )
+            # Compile each of the functions above, with an example input
+            # that has 5 in the first dimension, but is marked as dynamic
+
+            torch.compile(backend="inductor", dynamic=None)(fn_1)(_x)
+            torch.compile(backend="inductor", dynamic=None)(fn_2)(_x)
+
+            # @issue: If the line below is uncommented, an Exception is raised.
+            # torch.compile(backend="inductor", dynamic=None)(fn_3)(_x)
+        finally:
+            register_backend_for_device(
+                "cpu", orig_cpu_codegens.scheduling, orig_cpu_codegens.wrapper_codegen
+            )
+
 
-instantiate_device_type_tests(TestInductorDynamic, globals())
+# @TODO: Uncomment again
+# instantiate_device_type_tests(TestInductorDynamic, globals())
```

Comment:
> Add a test?

@jansel I have added one by now. I think it uncovered another preexisting issue that you might want to triage. If I uncomment this line, I get the following exception:

```
  File "/home/klondenberg/github/pytorch/pytorch/torch/_dynamo/guards.py", line 705, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/klondenberg/github/pytorch/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3308, in produce_guards
    raise ConstraintViolationError(
torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (L['x'].size()[0])! For more information, run with TORCH_LOGS="+dynamic".
  - Not all values of RelaxedUnspecConstraint(L['x'].size()[0]) are valid because L['x'].size()[0] was inferred to be a constant (5).
```

I think it's not a problem with this code. If I understand it all correctly, this Exception is triggered because a value that's marked to be dynamic is specialized to a static value in fn_3. But should this not just specialize and trigger recompilation if the guard is violated with a different input?

Reply:
This is expected, as you are asking for the dymension to be dynamic, and then it's specialised to be static. These `TestInductorDynamic` just runs the tests with dynamic shapes, but in this test you are handling the dynamic shapes yourself, so no need to instantiate them with `TestInductorDynamic`